documentation/Calls.md

```diff
@@ -0,0 +1,636 @@
+interface and that can possibly be referenced, directly or indirectly,
+with implicit interfaces are limited to argument lists that comprise
+only the addresses of effective arguments and the length of a `CHARACTER` function result
+(when there is one), and they can return only scalar values with constant
+type parameter values.
+None of their arguments or results need be (or can be) implemented
+with descriptors,
+and any internal procedures passed to them as arguments must be
+simple addresses of non-internal subprograms or trampolines for
+internal procedures.
+
+Note that `INTENT` and `CONTIGUOUS` attributes do not, by themselves,
+require the use of explicit interface; neither does the use of a dummy
+procedure (implicit or explicit in their interfaces).
+
+Analyses of calls to F77ish procedures must make
+allowances or impose restrictions capable of dealing with any of the
+invisible `INTENT` and `CONTIGUOUS` attributes of dummy arguments
+in the called procedure.
+
+## Protocol overview
+
+Here is a summary script of all of the actions that may need to be taken
+by the calling procedure and its referenced procedure to effect
+the call, entry, exit, and return steps of the procedure reference
+protocol.
+The order of these steps is not particularly strict, and we have
+some design alternatives that are explored further below.
+
+### Before the call:
+
+1. Compute &/or copy into temporary storage the values of
+   some effective argument expressions and designators (see below).
+1. Create and populate descriptors for arguments that use them
+   (see below).
+1. Possibly allocate function result storage,
+   when its size can be known by all callers; function results that are
+   neither `POINTER` nor `ALLOCATABLE` must have explicit shapes (C816).
+1. Create and populate a descriptor for the function result, if it
+   needs one (deferred-shape/-length `POINTER`, any `ALLOCATABLE`,
+   derived type with non-constant length parameters, &c.).
+1. Capture the values of host-escaping local objects in memory;
+   package them into single address (for calls to internal procedures &
+   for calls that pass internal procedures as arguments).
+1. Resolve the target procedure's polymorphic binding, if any.
+1. Marshal effective argument addresses (or values for `VALUE`/`%VAL()`) into registers.
+1. Marshal `CHARACTER` argument lengths in additional value arguments for
+   `CHARACTER` effective arguments not passed via descriptors.
+1. Marshal an extra argument for the length of a `CHARACTER` function
+   result if the function is F77ish.
+1. Marshal an extra argument for the function result's descriptor,
+   if it needs one.
+1. Set the "host instance" (static link) register when calling an internal
+   procedure from its host or another internal procedure, a procedure pointer,
+   or dummy procedure (when it has a descriptor).
+1. Jump.
+
+### On entry:
+1. Shuffle `ENTRY` dummy arguments & jump to common entry point.
```

Comment:
It is not really clear to me what is happening here I guess "`ENTRY` dummy arguments" means that the callee is a procedure with multiple `ENTRY` points. Is it already clear if we are going to have one compiled function per entry point or a single function with some kind of dispatcher at the beginning ?
In the first case, I do not really understand why we would need the set described here.
In the second case, wouldn't we need to pass some kind of index as arguments to indicate where the control flow should be dispatched ?